### DISCUSSION TITLE
for extended model undefined method App\Models\Role::hasRole()

### DISCUSSION THREAD

**Discussion by MehranLabour** [spatie/laravel-permission#1687] [Q&A] [answered]
my laravel version is: 6.2
my packege version is:    "spatie/laravel-permission": "^4.0",
i extended role model (`placed in :  App\Models\Role`) and the code below is extended model,
i used [this](https://spatie.be/docs/laravel-permission/v3/advanced-usage/extending) doc 
```
<?php

namespace App\Models;

use Spatie\Permission\Traits\HasPermissions;
use Spatie\Permission\Traits\RefreshesPermissionCache;

use Spatie\Permission\Contracts\Role As RoleContract;

use Spatie\Permission\Models\Role as SpatiteRole;

class Role extends SpatiteRole implements RoleContract
{

    use HasPermissions;
    use RefreshesPermissionCache;
 
    public function __construct(array $attributes = [])
    {
        parent::__construct($attributes);
        $this->fillable []= 'persian_name' ;
    }
    
}

```

and in config/permission.php i changed the default role model of packege like below code:
```
    'models' => [
        'role' => App\Models\Role::class,
    ],
```

am i did the steps right?

okay what is my problem?

 
assume  ` model_has_permissions` table is like below

```

 | permission_id | model_type | model_id |
 |            2          |   App/Role   |        1       |
 
```
'permissions'  table is like below
 
```
id  | name
1   | store 
2   | edit
3   | delete
```
my problem is when i want to use my new Role model like below code if Role model does  not have the permission throw an error

```
<?php

namespace App\Http\Controllers\Admin;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use App\Models\Role;
use App\Models\Permission;
use Spatie\Permission\Models\Role as SpatieRole;


class PermissionController extends Controller
{
    public function index()
    {
        $role=Role::Find(1);
        dd($role->hasPermissionTo('delete'));
    }
}
``` 

then below error will thrown 
```
BadMethodCallException
Call to undefined method App\Models\Role::hasRole()
```

but if i use SpatieRole model i mean the original model for finding role as below code 
```
<?php

namespace App\Http\Controllers\Admin;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use App\Models\Role;
use App\Models\Permission;
use Spatie\Permission\Models\Role as SpatieRole;


class PermissionController extends Controller
{
    public function index()
    {
        $role=SpatieRole::Find(1);
        dd($role->hasPermissionTo('delete'));
    }
}
``` 
nothing will thrown and result is `false`

i think some functions is not inherited by extended model,
i tried to explain my problem as good as i can,

any help will be useful

**Comment by drbyte** [comment, marked as answer]
> i extended role model (`placed in : App\Models\Role`) and the code below is extended model,
> i used [this](https://spatie.be/docs/laravel-permission/v3/advanced-usage/extending) doc
> 
> ```
> <?php
> 
> namespace App\Models;
> 
> use Spatie\Permission\Traits\HasPermissions;
> use Spatie\Permission\Traits\RefreshesPermissionCache;
> 
> use Spatie\Permission\Contracts\Role As RoleContract;
> 
> use Spatie\Permission\Models\Role as SpatiteRole;
> 
> class Role extends SpatiteRole implements RoleContract
> {
> 
>     use HasPermissions;
>     use RefreshesPermissionCache;
>  
>     public function __construct(array $attributes = [])
>     {
>         parent::__construct($attributes);
>         $this->fillable []= 'persian_name' ;
>     }
>     
> }
> ```

That seems overly complex.

Why not just:

```php
<?php

namespace App\Models;

use Spatie\Permission\Models\Role as SpatieRole;

class Role extends SpatieRole
{
    public function __construct(array $attributes = [])
    {
        parent::__construct($attributes);
        $this->fillable []= 'persian_name' ;
    }
}
```

**Reply by MehranLabour** [reply]
Ah,
As you said i did and ِ **YAAAY** !   problem solved
 i mixed up in documentation, more example in docs could be usefull  
Thanks So Much

**Reply by drbyte** [reply]
Docs updated (master branch)
